internal/common/tree.go

```diff
@@ -25,6 +26,173 @@ type DNSTreeNodeData struct {
 	Targets          []string
 }
 
+// PropagateStoppableLabel takes a propLabel (and value) to propagate throughout a tree, and a stopLabel
+// whenever the label is propagated to a dataset in a node which also has the stopLabel, this node and
+// all of the children of this node and any parents will have the propLabel removed.
+//
+// N.B. Any node with no parents is assumed to have the stopLabel, even when not present, to prevent the
+// propLabel propagating to the entire tree (if this is required, use `AddLabelToBranch` on the root node).
+//
+// The overview of the logic of this function is as follows:
+// - Spread propLabels as greedily as possible
+//   - Any labelled node labels all of it's children
+//   - Any node with all children labelled get's the propLabel too
+//
+// - Resolve the stopLabels
+//   - Any node with the stopLabel and the propLabel:
+//   - Has the propLabel removed from itself and all it's children
+//   - Has the propLabel removed from any parent (or parent's parent) that has the label
+
+func PropagateStoppableLabel(node *DNSTreeNode, propLabel, value, stopLabel string) {
+	//propagate labels regardless of stop labels
+	propagateLabel(node, propLabel, value)
+
+	//propagate stop labels
+	resolveStops(node, propLabel, value, stopLabel)
+}
+
+func resolveStops(node *DNSTreeNode, label, value, stopLabel string) {
+	if isRoot(node) && allChildrenHaveLabel(node, label, value) {
+		RemoveLabelFromTree(node, label)
+		//entire tree cleaned so no need for any further checks
+		return
+	}
+
+	//remove label from stop labelled children
+	for _, c := range node.Children {
+		d := findDataSetForChild(node, c.Name)
+		//has label and stop label = remove label from this dataset and all children
+		if d != nil && d.Labels[stopLabel] != "" && d.Labels[label] != "" {
+			delete(d.Labels, label)
+			RemoveLabelFromTree(c, label)
+			RemoveLabelFromParents(node, label)
```

Comment:
reading this made me go read RemoveLabelFromParents as I couldn't infer that it wouldn't remove it from the root node. Its a nit not a big deal. Maybe `RemoveLabelFromSubParents`?